librad/src/git/storage/fetcher.rs

```diff
@@ -262,6 +262,20 @@ pub mod error {
         #[error(transparent)]
         Pool(#[from] super::PoolError),
     }
+
+    #[derive(Debug, Error)]
+    pub enum FetchError {
+        #[error("Fetch limit {limit} exceeded, fetched: {amount_fetched} from {remote}")]
+        FetchLimitExceeded {
+            limit: usize,
+            amount_fetched: usize,
+            remote: PeerId,
+            fetch_specs: Fetchspecs<PeerId, Revision>,
+            refspecs: Vec<String>,
```

Comment:
Do we need `fetch_specs` _and_ `refspecs` out of curiosity? It seems `refspecs` can be derived from `fetch_specs`

Reply:
My inclination is to have both. Yes the `refspecs` are derived from the `fetchspecs`, but the logic is non-trivial. In my experience when things have gone wrong (which is when you would be looking at an error) you want to expose as much as possible of the intermediate state of the program so that it's easier to rule out hypotheses about which part of the program have misbehaved.

Reply:
Sounds reasonable to me :+1: